src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py

```diff
@@ -1,30 +1,154 @@
-#---------------------------------------------------------------------------------------------
+﻿#---------------------------------------------------------------------------------------------
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License. See License.txt in the project root for license information.
 #---------------------------------------------------------------------------------------------
 
 import uuid
-import re
 
 from azure.cli._util import CLIError
+from azure.cli.commands.client_factory import get_mgmt_service_client, configure_common_settings
+from azure.mgmt.authorization import AuthorizationManagementClient
+from azure.graphrbac import GraphRbacManagementClient
 from azure.mgmt.authorization.models import RoleAssignmentProperties
-from ._params import _auth_client_factory
+from azure.graphrbac.models import UserCreateParameters, PasswordProfile
+
+def _auth_client_factory(**_):
+    return get_mgmt_service_client(AuthorizationManagementClient)
+
+def _graph_client_factory(**_):
+    from azure.cli._profile import Profile
+    profile = Profile()
+    cred, _, tenant_id = profile.get_login_credentials(True)
+    client = GraphRbacManagementClient(cred, tenant_id)
+    configure_common_settings(client)
+    return client
 
-#TODO: expand the support to be in parity with node cli
-def create_role_assignment(role, object_id, scope=None):
+def create_role_assignment(role, assignee, resource_group_name=None, resource_id=None):
+    '''
+    :param assignee: represent a user, group, or service principal.
+    supported format: object id, user sign-in name, or service principal name
+    :param resource_id: resource id
+    '''
     assignments_client = _auth_client_factory().role_assignments
     definitions_client = _auth_client_factory().role_definitions
-    role_id = role
-    scope = scope or '/subscriptions/' + definitions_client.config.subscription_id
-    if not re.match(r'[0-9a-f]{32}\Z', role, re.I): #retrieve role id
+
+    if resource_id:
+        if resource_group_name:
+            err = 'Resource group "{}" is redundant because resource id is supplied'
+            raise CLIError(err.format(resource_group_name))
+        scope = resource_id
+    else:
+        scope = '/subscriptions/' + definitions_client.config.subscription_id
+        if resource_group_name:
+            scope = scope + '/resourceGroups/' + resource_group_name
+
+    role_id = None
+    try:
+        uuid.UUID(role)
+        role_id = role
+    except ValueError:
+        pass
```

Comment:
The preferred place to put argument validation is either in a validator (for validation that needs all parsed argument values) or in a type parameter (for individual argument values). This way, you get a consistent error message and usage statement instead of a single line of error text.